kopf/reactor/causation.py

```diff
@@ -0,0 +1,132 @@
+"""
+Detection of the event causes, based on the resource state.
+
+The low-level watch-events are highly limited in information on what
+caused them, and they only notify that the object was changed somehow:
+
+* ``ADDED`` for the newly created objects (or for the first-time listing).
+* ``MODIFIED`` for the changes of any field, be that metainfo, spec, or status.
+* ``DELETED`` for the actual deletion of the object post-factum.
+
+The conversion of the low-level *events* to the high-level *causes* is done by
+checking the object's state and comparing it to the saved last-seen state.
+
+This allows to track which specific fields were changed, and are those changes
+important enough to call the handlers: e.g. the ``status`` changes are ignored,
+so as some selected system fields of the ``metainfo``.
+
+For deletion, the cause is detected when the object is just marked for deletion,
+not when it is actually deleted (as the events notify): so that the handlers
+could execute on the yet-existing object (and its children, if created).
+"""
+import logging
+from typing import NamedTuple, Text, Mapping, MutableMapping, Optional, Any, Union
+
+from kopf.reactor import registries
+from kopf.structs import diffs
+from kopf.structs import finalizers
+from kopf.structs import lastseen
+
+# The constants for the event types, to prevent the direct string usage and typos.
```

Comment:
"Constants are used for event types, to prevent a direct usage of strings, and typos"